File: app/src/main/java/com/globalwarming/earthsaver/directories/DirectoryQuestionAdapter.java
package com.globalwarming.earthsaver.directories;

import android.text.Editable;
import android.text.TextWatcher;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.globalwarming.earthsaver.R;

import java.util.HashMap;
import java.util.List;

public class DirectoryQuestionAdapter extends RecyclerView.Adapter<DirectoryQuestionAdapter.ViewHolder> {

    private List<DirectoryQuestion> list;
    private HashMap<String, Pair<Boolean, String>> entries = new HashMap<>();

    public DirectoryQuestionAdapter(List<DirectoryQuestion> list) {
        this.list = list;
    }

    public HashMap<String, Pair<Boolean, String>> getEntries() {
        return entries;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_directory_question, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        DirectoryQuestion directoryQuestion = list.get(position);

        holder.textViewQuestion.setText(directoryQuestion.getQuestion());
        holder.textViewCategory.setText(directoryQuestion.getCategory());
        holder.textViewPoint.setText(directoryQuestion.getPoints() + " point");

        if (entries.containsKey(directoryQuestion.getId())) {
            holder.checkBox.setChecked(true);
            holder.editText.setVisibility(View.VISIBLE);
            holder.editText.setText(entries.get(directoryQuestion.getId()).second);
        } else  {
            holder.checkBox.setChecked(false);
            holder.editText.setVisibility(View.GONE);
        }

        holder.checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                holder.editText.setVisibility(View.VISIBLE);
                entries.put(directoryQuestion.getId(), Pair.create(true, ""));
            } else {
                holder.editText.setVisibility(View.GONE);
                entries.remove(directoryQuestion.getId());
            }
        });

        holder.editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                entries.put(directoryQuestion.getId(), Pair.create(entries.get(directoryQuestion.getId()).first, s.toString()));
            }
        });

    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        TextView textViewQuestion;
        TextView textViewCategory;
        TextView textViewPoint;
        CheckBox checkBox;
        EditText editText;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            textViewQuestion = itemView.findViewById(R.id.text_question);
            textViewCategory = itemView.findViewById(R.id.text_category);
            textViewPoint = itemView.findViewById(R.id.text_points);
            checkBox = itemView.findViewById(R.id.check_box);
            editText = itemView.findViewById(R.id.edit_text);

        }

    }

}

File: app/src/main/java/com/globalwarming/earthsaver/directories/DirectoryQuestionAdapter.kt
package com.globalwarming.earthsaver.directories

import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.globalwarming.earthsaver.databinding.LayoutDirectoryQuestionBinding
import com.google.android.material.snackbar.Snackbar
import java.util.*

class DirectoryQuestionAdapter : RecyclerView.Adapter<DirectoryQuestionAdapter.ViewHolder>() {

    private val list = ArrayList<DirectoryQuestion>()
    private val entries = HashMap<String, String>()

    fun setList(list: List<DirectoryQuestion>) {
        this.list.clear()
        this.list.addAll(list)
        notifyDataSetChanged()
    }

    fun getEntries(): HashMap<String, String> {
        return entries
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = LayoutDirectoryQuestionBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val directoryQuestion = list[position]
        holder.binding.textQuestion.text = directoryQuestion.question
        holder.binding.textCategory.text = directoryQuestion.category
        holder.binding.textPoints.text = "${directoryQuestion.points} point"
        holder.binding.checkBox.isEnabled = directoryQuestion.canRespond
        holder.binding.textAlreadyResponded.visibility = if (directoryQuestion.canRespond) {
            View.GONE
        } else {
            View.VISIBLE
        }
        if (entries.containsKey(directoryQuestion.id)) {
            holder.binding.checkBox.isChecked = true
            holder.binding.editText.visibility = View.VISIBLE
            holder.binding.editText.setText(entries[directoryQuestion.id])
        } else {
            holder.binding.checkBox.isChecked = false
            holder.binding.editText.visibility = View.GONE
        }
        holder.binding.checkBox.setOnCheckedChangeListener { checkBox, isChecked ->
            if (isChecked) {
                holder.binding.editText.visibility = View.VISIBLE
                entries[directoryQuestion.id] = ""
            } else {
                holder.binding.editText.visibility = View.GONE
                entries.remove(directoryQuestion.id)
            }
        }
        holder.binding.editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                entries[directoryQuestion.id] = s.toString()
            }
        })
        holder.binding.root.setOnClickListener {
            if (!directoryQuestion.canRespond) {
                Snackbar.make(
                    it,
                    "You have already engaged with this activity in the last 24 hours. Please try again in a day.",
                    Snackbar.LENGTH_SHORT
                ).show()
            }
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    class ViewHolder(val binding: LayoutDirectoryQuestionBinding) :
        RecyclerView.ViewHolder(binding.root)
}